### PR TITLE
Mutable matrix

### PR DESCRIPTION
**Pull request by marcandre**
This builds on #1769 from @grzegorz-jakubiak.

Problems found:

1) Debugging code left around. Simply use `make test-all TESTS='matrix/test_matrix.rb'` from the ruby source to test modified `lib/matrix.rb`

2) clone was not working. The existing `clone` probably confused you, it was wrong. Best is to specialize `initialize_copy`. Tests added.

3) Many errors with []=. See my extensive test that checks a big number of combinations of arguments:
* `2..-4` or `0...0` raised `NoMethodError` because `range.max #=> nil`
* `matrix[range, range] = Vector` not handled well
* `m[1..2][-2..3] = 42 #=> Matrix[[1, 2, 3, 4], [5, 6, 42, 42, 42, 42, 42, 42], [9, 10, 42, 42, 42, 42, 42, 42]]`
* `m[1..2][-2..3] = Matrix[[20, 21], [22, 23]] #=> erroneously raised ErrDimensionMismatch`
* `m[1..2][5] = 42 / Vector[42, 42] / Matrix[[42], [42]] # => Matrix[[1, 2, 3, 4], [5, 6, 7, 8, nil, 42], [9, 10, 11, 12, nil, 42]]`
* `m[±4][2..3] = 42  #=> raised NoMethodError: undefined method '[]=' for nil:NilClass`
* No handling of endless ranges (those are new to Ruby 2.6)

4) Improved error messages

I still have to go over the changes to Vector.

### CODE REVIEW

**Comment by grzegorz-jakubiak** [comment]
Thanks for working on this. I hope there's less work to do with Vector. 😄

**Comment by grzegorz-jakubiak** [comment]
How's going? I've heard that `matrix` has been promoted to default gem. Does this change anything in adding this feature? I think we've forgotten about making `[]` method accept range since we made `[]=` accept range and I think the parameters for retrieving should be the same 😉

**Comment by marcandre** [comment]
> I've heard that matrix has been promoted to default gem.

Right, most (all?) standard libraries are being gemified. AFAIK, this won't change anything for Ruby 2.5 or earlier.

> Does this change anything in adding this feature?

Not really, no.

> I think we've forgotten about making [] method accept range

Indeed, agreed. Best make a separate PR.

**Comment by grzegorz-jakubiak** [comment]
Okay I see. How's progress with `Vector`'s changes? 😉  Regarding PR if I wanted to do it should I submit it here or `ruby/matrix`?

**Comment by marcandre** [comment]
Got time to check out Vector. Fixed `clone` and tweaked some tests, but otherwise all good.

@grzegorz-jakubiak Is it possible for you to check this PR before I squash and commit this?

File: lib/matrix.rb [hunk 1]
```diff
@@ -303,12 +303,107 @@ def [](i, j)
...
+      set_value(row, column, v)
+    end
   end
   alias set_element []=
   alias set_component []=
-  private :[]=, :set_element, :set_component
+  private :set_element, :set_component
+
+  # Returns range or nil
+  private def check_range(val, direction)
```
Comment (grzegorz-jakubiak) [inline lib/matrix.rb]: Is it a new convention to use `private` keyword inline?

File: lib/matrix.rb [hunk 2]
```diff
@@ -1872,6 +2055,20 @@ def zero?
     all?(&:zero?)
   end
 
+  def freeze
+    @elements.freeze
+    super
```
Comment (grzegorz-jakubiak) [inline lib/matrix.rb]: Same here

File: lib/matrix.rb [hunk 3]
```diff
@@ -361,17 +456,49 @@ def column(j) # :yield: e
...
+  def collect!(which = :all)
+    return to_enum(:collect!, which) unless block_given?
+    raise FrozenError, "can't modify frozen Matrix" if frozen?
+    each_with_index(which){ |e, row_index, col_index| @rows[row_index][col_index] = yield e }
+  end
+
+  alias map! collect!
+
+  def freeze
+    @rows.freeze
```
Comment (grzegorz-jakubiak) [inline lib/matrix.rb]: Shouldn't super be called first?

File: lib/matrix.rb (hunk 1 above)
Comment (marcandre) [reply lib/matrix.rb]: Well, it's *my* new convention :-)

File: lib/matrix.rb (hunk 3 above)
Comment (marcandre) [reply lib/matrix.rb]: It could, in this case. Any difference?

Comment (marcandre) [reply lib/matrix.rb]: :+1:
In general, I'd say that `freeze` is a case where you want to call super at the end, for example if you have local cache, you need to do `@cache ||= ...` before `super`.

Comment (grzegorz-jakubiak) [reply lib/matrix.rb]: Right I didn't think of it at first 😉

**Comment by grzegorz-jakubiak** [comment]
@marcandre just reminding you of this in case you forgot 😉

**Comment by marcandre** [comment]
:-) Sure!

Unless I'm mistaken, remains to be done:
1) Handling endless range
2) Handling `Matrix#[]` and `Vector#[]` to

**Comment by marcandre** [comment]
> I thought you've already handled endless ranges?

Indeed, I had forgotten. So only `Matrix#[]` and `Vector#[]` remain then.

> Do you think this can still be merged before 2.6 is released?

Of course. release is at X-mas time.

**Comment by grzegorz-jakubiak** [comment]
I think `Vector#[]` is actually done as well cause `Array#[]` already supports it. I can't push to your branch so I pushed to mine: https://github.com/ruby/ruby/pull/1769/commits/76eb471b373d7756634025242f28750e5e0881d0. Have a look and `cherry-pick` it if you want, hope this makes your work on that easier 😉 . If I've got more time, I'll try to work on `Matrix#[]`.

> Of course. release is at X-mas time.

Awesome! Can't wait!

**Pull request closed without merging**